x/incentive/keeper/rewards.go

```diff
@@ -0,0 +1,110 @@
+func (k Keeper) CreateNewRewardPeriod(ctx sdk.Context, reward types.Reward) {
+	// reward periods store the amount of rewards payed PER SECOND
+	rewardsPerSecond := sdk.NewDecFromInt(reward.Reward.Amount).Quo(sdk.NewDecFromInt(sdk.NewInt(int64(reward.Duration.Seconds())))).TruncateInt()
+	rewardCoinPerSecond := sdk.NewCoin(reward.Reward.Denom, rewardsPerSecond)
+	rp := types.RewardPeriod{
+		Denom:         reward.Denom,
+		Start:         ctx.BlockTime(),
+		End:           ctx.BlockTime().Add(reward.Duration),
+		Reward:        rewardCoinPerSecond,
+		ClaimEnd:      ctx.BlockTime().Add(reward.Duration).Add(reward.ClaimDuration),
+		ClaimTimeLock: reward.TimeLock,
+	}
+	k.SetRewardPeriod(ctx, rp)
+}
+
+// CreateAndDeleteRewardPeriods creates reward periods for active rewards that don't already have a reward period and deletes reward periods for inactive rewards that currently have a reward period
+func (k Keeper) CreateAndDeleteRewardPeriods(ctx sdk.Context) {
+	params := k.GetParams(ctx)
+
+	for _, r := range params.Rewards {
+		_, found := k.GetRewardPeriod(ctx, r.Denom)
+		// if governance has made a reward inactive, delete the current period
+		if !r.Active {
+			k.DeleteRewardPeriod(ctx, r.Denom)
+		}
+		// if a reward period for an active reward is not found, create one
+		if !found && r.Active {
+			k.CreateNewRewardPeriod(ctx, r)
+		}
+	}
+}
+
+// ApplyRewardsToCdps iterates over the reward periods and creates a claim for each cdp owner that created usdx with the collateral specified in the reward period
+func (k Keeper) ApplyRewardsToCdps(ctx sdk.Context) {
+	previousBlockTime, found := k.GetPreviousBlockTime(ctx)
+	if !found {
+		previousBlockTime = ctx.BlockTime()
+		k.SetPreviousBlockTime(ctx, previousBlockTime)
+		return
+	}
+	k.IterateRewardPeriods(ctx, func(rp types.RewardPeriod) bool {
+		expired := false
+		// the total amount of usdx created with the collateral type being incentivized
+		totalPrincipal := k.cdpKeeper.GetTotalPrincipal(ctx, rp.Denom, types.PrincipalDenom)
+		// the number of seconds since last payout
+		timeElapsed := sdk.NewInt(ctx.BlockTime().Unix() - previousBlockTime.Unix())
+		if rp.End.Before(ctx.BlockTime()) {
+			timeElapsed = sdk.NewInt(rp.End.Unix() - previousBlockTime.Unix())
+			expired = true
+		}
+		// the amount of rewards to pay (rewardAmount * timeElapsed)
+		rewardsThisPeriod := rp.Reward.Amount.Mul(timeElapsed)
+		id := k.GetNextClaimPeriodID(ctx, rp.Denom)
+		k.cdpKeeper.IterateCdpsByDenom(ctx, rp.Denom, func(cdp cdptypes.CDP) bool {
+			rewardsShare := sdk.NewDecFromInt(cdp.Principal.AmountOf(types.PrincipalDenom).Add(cdp.AccumulatedFees.AmountOf(types.PrincipalDenom))).Quo(sdk.NewDecFromInt(totalPrincipal))
+			// sanity check - don't create zero claims
+			if rewardsShare.IsZero() {
+				return false
+			}
+			rewardsEarned := rewardsShare.Mul(sdk.NewDecFromInt(rewardsThisPeriod)).RoundInt()
+			k.AddToClaim(ctx, cdp.Owner, types.GovDenom, id, sdk.NewCoin(types.GovDenom, rewardsEarned))
```

Comment:
```suggestion
			k.AddToClaim(ctx, cdp.Owner, rp.Denom, id, sdk.NewCoin(types.GovDenom, rewardsEarned))
```

Shouldn't this be the collateral/input denom as opposed to gov denom?

The test suite has `suite.keeper.AddToClaim(suite.ctx, addr, "xpr", 1, c("ukava", 1000000))` so manually added claims are probably working. But this method creates claims under the reward/output denom instead of the collateral/input denom, so they can't be properly claimed.